ncm-systemd/src/main/perl/Systemd/UnitFile.pm

```diff
@@ -0,0 +1,444 @@
+{
+    my ($self) = @_;
+
+    if (!(blessed($self->{config}) &&
+           $self->{config}->isa("EDG::WP4::CCM::Element"))) {
+        $self->error("config has to be an Element instance");
+        return;
+    }
+
+    # custom values
+    my $custom = $self->custom();
+    return if (! defined($custom));
+
+    # prepare/cleanup destination and return filename
+    my $filename = $self->_prepare_path($UNITFILE_DIRECTORY);
+    return if(! defined($filename));
+
+    # render
+    my $trd = EDG::WP4::CCM::TextRender->new(
+        $UNITFILE_TT,
+        $self->{config},
+        relpath => 'systemd',
+        log => $self,
+        ttoptions => _make_variables_custom($custom),
+        );
+
+    # write
+    my $fh = $trd->filewriter(
+        $filename,
+        backup => $self->{backup},
+        mode => 0664,
+        log => $self,
+        );
+
+    if(! defined($fh)) {
+        $self->error("Rendering unitfile for unit $self->{unit}",
+                     " (filename $filename) failed: $trd->{fail}.");
+        return;
+    }
+
+    my $changed = $fh->close() ? 1 : 0; # force to 1 or 0
+
+    # if changed, reload daemon
+    if($changed) {
+        # can't do much with return value?
+        CAF::Process->new(
+            [$SYSTEMCTL, $DAEMON_RELOAD],
+            log => $self,
+            )->execute();
+    }
+
+    return $changed;
+}
+
+=pod
+
+=back
+
+=head2 Private methods
+
+=over
+
+=item _prepare_path
+
+Create and return the filename to use,
+and prepare the directory structure if needed.
+
+C<basedir> is the base directory to use, e.g. C<$UNITFILE_DIRECTORY>.
+
+=cut
+
+sub _prepare_path
+{
+    my ($self, $basedir) = @_;
+
+    my $filename;
+
+    my $unitfile = "$basedir/$self->{unit}";
+    my $unitdir = "${unitfile}.d";
+
+    if ($self->{replace}) {
+        # unitdir can't exist
+        return if (! $self->_cleanup($unitdir));
+
+        $filename = $unitfile;
+    } else {
+        # unitfile can't exist
+        return if (! $self->_cleanup($unitfile));
+
+        $filename = "$unitdir/$NOREPLACE_FILENAME";
+        if (! ($self->_directory_exists($unitdir) || $self->_make_directory($unitdir))) {
+            $self->error("Failed to create unitdir $unitdir: $!");
+            return;
+        }
+    };
+
+    return $filename;
+}
+
+=item _hwloc_calc_cpuaffinity
+
+Run C<_hwloc_calc_cpus>, and returns in C<CPUAffinity> format with a reset
+
+=cut
+
+sub _hwloc_calc_cpuaffinity
+{
+    my ($self, $locations) = @_;
+
+    my $cpus = $self->_hwloc_calc_cpus($locations);
+    return if(! defined($cpus));
+
+    # first empty list, to reset all previous defined CPUaffinity settings
+    return [[], $cpus];
+}
+
+
+=item _hwloc_calc_cpus
+
+Run the C<hwloc-calc --physical --intersect PU> command for C<locations>.
+
+Returns arrayref with CPU indices on success, undef otherwise.
+
+=cut
+
+sub _hwloc_calc_cpus
+{
+    my ($self, $locations) = @_;
+
+    local $ENV{HWLOC_HIDE_ERRORS};
+    $ENV{HWLOC_HIDE_ERRORS}=1;
```

Comment:
good point. when your bios cpu mapping doens't make much sense to hwloc, it will print a nice header mentioning this, and still try to guess the correct cpu ids. ofcourse it uses stdout for this instead of stderr (well, at least with the version i saw this).
i could try to report this, logging any output not matching the `^\d+(,\d+)*$`. shouldn't be that much work. unfornately, we fixed all our bioses, so i have no machines anymore that spit out the messages :smile: